modules/onboarding/src/main/kotlin/edu/stanford/spezi/module/onboarding/spezi/flow/OnboardingStack.kt

```diff
@@ -0,0 +1,105 @@
+package edu.stanford.spezi.module.onboarding.spezi.flow
+
+import androidx.compose.foundation.layout.Column
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.compositionLocalOf
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Modifier
+import androidx.navigation.compose.NavHost
+import androidx.navigation.compose.rememberNavController
+import androidx.navigation.createGraph
+import edu.stanford.spezi.core.design.component.Button
+import edu.stanford.spezi.core.design.theme.SpeziTheme
+import edu.stanford.spezi.core.design.theme.ThemePreviews
+
+internal data class OnboardingNavigationStep(
+    val id: String,
+    val content: @Composable () -> Unit,
+)
+
+val LocalOnboardingNavigationPath = compositionLocalOf<OnboardingNavigationPath> {
+    error("Do not access this outside of OnboardingStack!")
+}
+
+@Composable
+fun OnboardingStack(
```

Comment:
Would propose the following here and usage of rembemberSaveable as that is what also rember nav controller useses:

```kotlin
@Composable
fun OnboardingStack(
    modifier: Modifier = Modifier,
    onComplete: () -> Unit = {},
    startAtStep: String? = null,
    content: OnboardingStackBuilder.() -> Unit,
) {
    val navController = rememberNavController()
    val navigationPath = rememberSaveable(navController, startAtStep, content) {
        OnboardingNavigationPath(navController, startAtStep, content)
    }

    navigationPath.NavHost(modifier = modifier)

    LaunchedEffect(navController) {
        navController.currentBackStackEntryFlow.collect { entry ->
            if (entry.destination.route == null) {
                onComplete()
            }
        }
    }
}
```

As per navigation path:

```kotlin
class OnboardingNavigationPath internal constructor(
    private val navController: NavHostController,
    startDestination: String?,
    builder: OnboardingStackBuilder.() -> Unit,
) {
    private val steps = OnboardingStackBuilder().apply(builder).steps.also {
        if (it.isEmpty()) error("No steps provided")
    }
    private val startDestination = startDestination ?: steps.first().id
    // rest of the content
    
    // nav host
    @Composable
    fun NavHost(modifier: Modifier) {
        CompositionLocalProvider(LocalOnboardingNavigationPath provides this) {
            NavHost(
                modifier = modifier,
                navController = navController,
                startDestination = startDestination,
            ) {
                composable(
                    route = customRoute("{id}"),
                    arguments = listOf(navArgument("id") { type = NavType.StringType })
                ) { backStackEntry ->
                    val step = backStackEntry.arguments?.getString("id")?.let { stepId ->
                        customSteps.firstOrNull { it.id == stepId }
                    }
                    step?.content?.let { it() } ?: IllegalOnboardingStep()
                }

                for (step in steps) {
                    composable(step.id) { step.content() }
                }
            }
        }
    }
```

I just replicated your logic in the custom route, but I personally see that logic not being needed